modules/ssh/agent.go

```diff
@@ -0,0 +1,112 @@
+package ssh
+
+import (
+	"crypto/x509"
+	"encoding/pem"
+	"io"
+	"io/ioutil"
+	"net"
+	"os"
+	"path/filepath"
+	"testing"
+
+	"golang.org/x/crypto/ssh/agent"
+)
+
+type SshAgent struct {
+	stop       chan bool
+	stopped    chan bool
+	socketDir  string
+	socketFile string
+	agent      agent.Agent
+	ln         net.Listener
+}
+
+// Create SSH agent, start it in background and returns control back to the main thread
+func NewSshAgent(t *testing.T, socketDir string, socketFile string) (*SshAgent, error) {
+	var err error
+	s := &SshAgent{make(chan bool), make(chan bool), socketDir, socketFile, agent.NewKeyring(), nil}
+	s.ln, err = net.Listen("unix", s.socketFile)
+	if err != nil {
+		return nil, err
+	}
+	go s.run(t)
+	return s, nil
+}
+
+// SSH Agent listener and handler
+func (s *SshAgent) run(t *testing.T) {
+	defer close(s.stopped)
+	for {
+		select {
+		case <-s.stop:
+			return
+		default:
+			c, err := s.ln.Accept()
+			if err != nil {
+				select {
+				// When s.Stop() closes the listener, s.ln.Accept() returns an error that can be ignored
+				// since the agent is in stopping process
+				case <-s.stop:
+					return
+					// When s.ln.Accept() returns a legit error, we print it and continue accepting further requests
+				default:
+					t.Logf("could not accept connection to agent %v", err)
+					continue
+				}
+			} else {
+				defer c.Close()
+				go func(c io.ReadWriter) {
+					err := agent.ServeAgent(s.agent, c)
+					if err != nil {
+						t.Logf("could not serve ssh agent %v", err)
+					}
+				}(c)
+			}
+		}
+	}
+}
+
+// Stop and clean up SSH agent
+func (s *SshAgent) Stop() {
+	close(s.stop)
+	s.ln.Close()
+	<-s.stopped
+	os.RemoveAll(s.socketDir)
+}
+
+// Instantiates and returns an in-memory ssh agent with the given KeyPair already added
+func SshAgentWithKeyPair(t *testing.T, keyPair *KeyPair) *SshAgent {
+	return SshAgentWithKeyPairs(t, []*KeyPair{keyPair})
+}
+
+// Instantiates and returns an in-memory ssh agent with the given KeyPair(s) already added
+func SshAgentWithKeyPairs(t *testing.T, keyPairs []*KeyPair) *SshAgent {
+	t.Log("Generating SSH Agent with given KeyPair(s)")
+
+	// Instantiate a temporary SSH agent
+	socketDir, err := ioutil.TempDir("", "ssh-agent-")
+	if err != nil {
+		t.Fatal(err)
+	}
+	socketFile := filepath.Join(socketDir, "ssh_auth.sock")
+	os.Setenv("SSH_AUTH_SOCK", socketFile)
```

Comment:
Hm, this won't work with multiple tests running in parallel. Is it strictly necessary to set this env var?

Reply:
I just copied this from the existing example, but I just tested with that line commented out and it fails to connect. Will look into this further to see if there's away around it.

Reply:
Here is what requires the env variable: https://github.com/gruntwork-io/terratest/blob/master/modules/ssh/ssh.go#L328

Question -- does it ever make sense to read from the local ssh-agent, when `Host.SshAgent == true`? In my opinion they shouldn't do that, because it makes the tests brittle and not very portable. They'll only work if the key(s) are present in the agent of the machine running the tests. 

Either way I think the solution is to store a reference to the instance of `SshAgent` on the Host struct:

	privateHost := ssh.Host{
		Hostname:    privateInstanceIP,
		SshUserName: "ubuntu",
		OverrideSshAgent: sshAgent,  // or replace the existing SshAgent bool?
	}

Then you could add another if statement (or replace the SshAgent bool) here: https://github.com/gruntwork-io/terratest/blob/master/modules/ssh/ssh.go#L327

(Replacing the bool would probably be a backwards incompatible change.)

Reply:
Ah, good find. Adding an `OverrideSshAgent` to `ssh.Host` sounds like a clean solution 👍

Reply:
Ok.. updated branch with the `OverrideSshAgent` change. I also just added another unit test which confirms that it can run multiple agents at once without mixing up keys. One thing to note is that I refactored the conditionals in `createAuthMethodsForHost` (they were getting unwieldy and confusing). As far as I can tell, the behavior is the same, but a second set of eyes couldn't hurt.